docs/style-guide.rst

```diff
@@ -0,0 +1,274 @@
+
+Strings
+-------
+
+Strings substitutions **should** be performed via `formatted string literals <https://docs.python.org/3/reference/lexical_analysis.html#formatted-string-literals>`_ rather than the ``str.format`` method or other techniques.
+
+Type Annotations
+----------------
+
+    * All publicly exposed classes and methods **should** include `PEP 484 <https://www.python.org/dev/peps/pep-0484/>`_ annotations for all arguments and return values.
+    * Type annotations **should** be included directly in the source. `Stub files <https://www.python.org/dev/peps/pep-0484/#stub-files>`_ **may** be used where there is a valid reason. Source files using stubs **must** still be annotated to aid readability.
+    * Internal methods **should** include type annotations.
+
+Tests
+=====
+
+We use the `pytest <https://docs.pytest.org/en/latest/>`_ framework for testing, and :ref:`eth-tester<testing-contracts-ethtester>` for our local development chain.
+
+Best Practices
+--------------
+
+    * ``pytest`` functionality **should not** be imported with ``from ...`` style syntax, particularly :func:`pytest.raises<pytest.raises>`. Importing the library itself aids readability.
+    * Tests **must not** be interdependent. We use ``xdist`` to execute tests in parallel. You **cannot** rely on which order tests will execute in, or that two tests will execute in the same process.
+    * Test cases **should** be designed with a minimalistic approach. Each test should verify a single behavior. A good test is one with few assertions, and where it is immediately obvious exactly what is being tested.
+    * Where logical, tests **should** be `parametrized <https://docs.pytest.org/en/latest/parametrize.html>`_ or use `property-based <https://hypothesis.works/>`_ testing.
+    * Tests **must not** involve mocking.
+
+
+Directory Structure
+-------------------
+
+Where possible, the test suite **should** copy the structure of main Vyper package. For example, test cases for ``vyper/context/types/`` should exist at ``tests/context/types/``.
+
+Filenames
+---------
+
+Test files **must** use the following naming conventions:
+
+    * ``test_[module].py``: When all tests for a module are contained in a single file.
+    * ``test_[module]_[functionality].py``: When tests for a module are split across multiple files.
+
+Fixtures
+--------
+
+    * Fixtures **should** be stored in ``conftest.py`` rather than the test file itself.
+    * ``conftest.py`` files **must not** exist more than one subdirectory beyond the initial ``tests/`` directory.
+    * The functionality of a fixture **must** be fully documented, either via docstrings or comments.
+
+Documentation
+=============
+
+It is important to maintain comprehensive and up-to-date documentation for the Vyper language.
+
+    * Documentation **must** accurately reflect the current state of the master branch on Github.
+    * New functionality **must not** be added without corresponding documentation updates.
```

Comment:
NOTE: All VIPs should have an accompanying documentation update and add new/modify tests in some way

Reply:
I think explaining the VIP process would fit better in the contribution guide than the style guide.  Right now our contribution guide looks like [this](https://vyper.readthedocs.io/en/latest/contributing.html), which is not bad but could definitely be expanded upon.  Also we should add a `CONTRIBUTING.md` that links to the docs so it's easier to find from the repo.

I'm going to open a new issue about this.

Reply:
#1905